### PR TITLE
Optimizations of Dilithium's signing procedure

### PR DESCRIPTION
**Pull request by PRASANNA-RAVI**

### CODE REVIEW

**Comment by PRASANNA-RAVI** [comment]
Hi Matthias,
        I had seen that I had updated an old implementation. Sorry about the same. I just finished adding my new changes and I have created a new pull request... In the dilithium2, dilithium3 and dilithium4, the updated implementations are there in the opt2 folder. Please let me know if they are okay.

With Thanks and Regards,
Prasanna Ravi.

File: Makefile [hunk 1]
```diff
@@ -39,8 +39,8 @@ RANDOMBYTES_M4=common/randombytes.c
 DEST_HOST=bin-host
 DEST=bin
 
-TARGET_NAME = $(shell echo $(IMPLEMENTATION_PATH) | sed 's@/@_@g')
-TYPE = $(shell echo $(IMPLEMENTATION_PATH) | sed -r 's@(.*/)?crypto_(kem|sign)(/.*)@\2@')
+TARGET_NAME = $(shell echo $(IMPLEMENTATION_PATH) | gsed 's@/@_@g')
+TYPE = $(shell echo $(IMPLEMENTATION_PATH) | gsed -r 's@(.*/)?crypto_(kem|sign)(/.*)@\2@')
```
Comment (Ko-) [inline Makefile]: Was there a reason to change this? `gsed` is not a standard command. I don't have it and I'm not sure what it does. If it just enables `global` substitution, then that's what the `g` in `s@/@_@g` already does.

File: crypto_sign/dilithium2/opt/sign.c [hunk 2]
```diff
@@ -6,7 +6,7 @@
 #include "randombytes.h"
 #include "sign.h"
 #include "symmetric.h"
-
+#include "hal.h"
```
Comment (Ko-) [inline crypto_sign/dilithium2/opt/sign.c]: This include is not necessary or desired, right? Was there a reason for adding it?

File: crypto_sign/dilithium2/opt/sign.c [hunk 3]
```diff
@@ -42,7 +42,8 @@ void expand_mat(polyvecl mat[K], const unsigned char rho[SEEDBYTES]) {
 **************************************************/
 void challenge(poly *c,
                                         const unsigned char mu[CRHBYTES],
-                                        const polyveck *w1) {
+                                        const polyveck *w1)
+                                        {
```
Comment (Ko-) [inline crypto_sign/dilithium2/opt/sign.c]: Changes like this one make it a bit harder to see what actually changed... I guess this was some IDE that liked forcing a particular code style?

File: crypto_sign/dilithium3/opt/sign.c [hunk 4]
```diff
@@ -6,7 +6,7 @@
 #include "randombytes.h"
 #include "sign.h"
 #include "symmetric.h"
-
+#include "hal.h"
```
Comment (Ko-) [inline crypto_sign/dilithium3/opt/sign.c]: Idem, probably not necessary.

File: crypto_sign/dilithium4/opt/sign.c [hunk 5]
```diff
@@ -6,7 +6,7 @@
 #include "randombytes.h"
 #include "sign.h"
 #include "symmetric.h"
-
+#include "hal.h"
```
Comment (Ko-) [inline crypto_sign/dilithium4/opt/sign.c]: Idem, probably not necessary.

File: Makefile (hunk 1 above)
Comment (thomwiggers) [reply Makefile]: It's probably GNU sed, it's possible that you're using some GNU extensions that are not in BSD (MacOS) sed

**Comment by Ko-** [comment]
If there's assembly in there (and there is: `ntt_asm.s`), this should not be called `opt`, but it can remain `m4`. Also, I don't see any `opt2` directory.

File: Makefile (hunk 1 above)
Comment (Ko-) [reply Makefile]: Ah yeah you're right, thanks. But I don't think that this line uses any GNU extensions. If I execute these lines with `--posix` which disables all GNU extensions it still works. I don't have a BSD/MacOS system to test this.

Comment (thomwiggers) [reply Makefile]: I don't think `-r` works on POSIX sed

> (both versions): use only the -n and -e options (notably, do not use -E or -r to turn on support for extended regular expressions)

> GNU sed: add option `--posix` to ensure POSIX-only functionality (you don't strictly need this, but without it you could end up inadvertently using non-POSIX features without noticing; caveat: `--posix` itself is not POSIX-compliant)

https://riptutorial.com/sed/topic/9436/bsd-macos-sed-vs--gnu-sed-vs--the-posix-sed-specification

Comment (Ko-) [reply Makefile]: I just found the same answer at https://unix.stackexchange.com/questions/145402/regex-alternation-or-operator-foobar-in-gnu-or-bsd-sed/145409.

File: crypto_sign/dilithium2/opt/sign.c (hunk 2 above)
Comment (PRASANNA-RAVI) [reply crypto_sign/dilithium2/opt/sign.c]: I had included it for my earlier project to add some gpio functions...But, I have removed the hal include now in the latest commit...

File: crypto_sign/dilithium2/opt/sign.c (hunk 3 above)
Comment (PRASANNA-RAVI) [reply crypto_sign/dilithium2/opt/sign.c]: That was a typo...sorry abt that!! I have changed it

File: crypto_sign/dilithium3/opt/sign.c (hunk 4 above)
Comment (PRASANNA-RAVI) [reply crypto_sign/dilithium3/opt/sign.c]: I had included it for my earlier project to add some gpio functions...But, I have removed the hal include now in the latest commit...

File: crypto_sign/dilithium4/opt/sign.c (hunk 5 above)
Comment (PRASANNA-RAVI) [reply crypto_sign/dilithium4/opt/sign.c]: I had included it for my earlier project to add some gpio functions...But, I have removed the hal include now in the latest commit...

File: Makefile (hunk 1 above)
Comment (PRASANNA-RAVI) [reply Makefile]: I had used gsed as sed didnt work on my mac OS...

Comment (thomwiggers) [reply Makefile]: You can probably install https://formulae.brew.sh/formula/gnu-sed

File: crypto_kem/r5nd-5kemcca-5d-sneik/m4/ct_util.c [hunk 6]
```diff
@@ -1 +0,0 @@
-../../r5nd-1kemcca-5d/m4/ct_util.c
```
Comment (thomwiggers) [inline crypto_kem/r5nd-5kemcca-5d-sneik/m4/ct_util.c]: You seem to be symlinks by the actual files, in an unrelated scheme?

File: Makefile (hunk 1 above)
Comment (Ko-) [reply Makefile]: I think this one should work: `sed 's@\([a-z_-]*/\)*crypto_\([a-z]*\).*@\2@'`. That's pure POSIX BRE without any option.

File: crypto_kem/r5nd-5kemcca-5d-sneik/m4/ct_util.c (hunk 6 above)
Comment (Ko-) [reply crypto_kem/r5nd-5kemcca-5d-sneik/m4/ct_util.c]: They were symlinks (which is fine in `pqm4`), but I think they were now changed accidentally into regular files. In any case a Dilithium PR should preferably not touch Round5 code.

File: Makefile (hunk 1 above)
Comment (PRASANNA-RAVI) [reply Makefile]: I have reverted it back to sed...

Comment (PRASANNA-RAVI) [reply Makefile]: Thanks for the inputs!!!

File: crypto_kem/r5nd-5kemcca-5d-sneik/m4/ct_util.c (hunk 6 above)
Comment (PRASANNA-RAVI) [reply crypto_kem/r5nd-5kemcca-5d-sneik/m4/ct_util.c]: I have not changed anything in the crypto_kem folder... May I know how I can resolve this issue??

Comment (PRASANNA-RAVI) [reply crypto_kem/r5nd-5kemcca-5d-sneik/m4/ct_util.c]: I tried running the test_no_symlinks.py in the pqclean folder and it returns "OK"...

Comment (Ko-) [reply crypto_kem/r5nd-5kemcca-5d-sneik/m4/ct_util.c]: No problem, I fixed that in 557ea5a.

**Comment by mkannwischer** [comment]
Thanks for the PR, Prasanna. 

When running the tests for dilithium4 (`python3 test.py dilithium4`) I'm getting

```
crypto_sign_keypair DONE.

crypto_sign DONE.

ERROR Signature did not verify correctly!
```

For dilithium2 and dilithium3 it seems to work. 
Can you reproduce this?

**Comment by PRASANNA-RAVI** [comment]
Hi Matthias,
               I also checked at my end. I am getting the same error at my end as well... :( Sorry, I had overlooked the issue. I had earlier checked correctness with Dilithium2 and assumed would be the same with all the parameter sets since only the size of the module changes.

Strange that it is working with dilithium2,3 and not 4... I think it might be some runtime issue. I ran the same code on my desktop and there does not seem to be any issue there for dilithium4 and all the parameter sets. Moreover, I removed the inlining and unrolling optimizations and only kept the optimizations in the sign.c file and it seems to be working fine. So, I think the issue is with one of the functions which was inlined or unrolled in poly.c.

I will look into it and resolve it as soon as possible. Thanks for the notification and sorry for the inconvenience caused.  It would be great if you can also notify me if have time to look into the issue.

With Thanks and Regards,
Prasanna Ravi.
________________________________

**Comment by PRASANNA-RAVI** [comment]
Hi Matthias,
                   I quickly did some tests at my end. I could see that the optimized implementation of dilithium4 runs correctly on the desktop, but not on the ARM Cortex-M4. I also checked the stack usage and it is the same for both the current m4 implementation and my implementation. So, I guess it is not an issue of stack overflow or any other memory error. I also tried different compilers but in vain. So, probably the issue is present somewhere at the lower level. I am currently caught up with another work whose deadline is fast approaching. So, I will get back to fix this issue in probably after a week's time. The mistake is on my part, since I had accidentally overlooked the correctness of dilithium4, since the modifications (in sign.c and poly.c) are identical for all the parameter sets.

So, you can let me know if I can keep the pull request open for now, or should I close it and repoen it again?

Since dilithium2 and dilithium3 are working, will you integrate them into the library while I fix dilithium4 and create a new pull request for the same?

Please reply at your earliest convenience.

With Thanks and Regards,
Prasanna Ravi.
________________________________

**Comment by PRASANNA-RAVI** [comment]
Hello,
        I think I have resolved the issue. The implementation of dilithium4 is also working fine and I have pushed all the changes to my master branch. Please let me know if there are any more issues.

With Thanks and Regards,
Prasanna Ravi.

**Comment by mkannwischer** [comment]
Thanks, I can confirm it works now. I've benchmarked it. Do those numbers roughly align with what you expected? 
There are a bunch of unnecessary changes like changing whitespace, but I'll merge this nonetheless.

**Comment by mkannwischer** [comment]
Note that this implementation now combines the tricks from 
https://eprint.iacr.org/2019/420 
with the NTT from 
https://www.emsec.ruhr-uni-bochum.de/media/seceng/veroeffentlichungen/2018/10/17/paper.pdf

**Comment by PRASANNA-RAVI** [comment]
Dear Matthias,
              Thanks for the confirmation. Actually, the improvements are significant and have meaning only when observing over a large number of measurements. Because the repetition rate of the signing procedure is pretty random. Moreover, the optimization mainly works by rejecting an iteration at the earliest possible instance. Hence, it reduces the time in rejected iterations. Thus, more the repetition rate, more the improvement you get.

When the first coefficient of a module is faulty, you get great improvement but when the last coefficient is faulty, you don't get much. So, only by measuring over a large number of signatures, you can observe a considerable improvement in the AVERAGE performance of the signing procedure. Hence, for the desktop implementation we got numbers for 10^7 signatures and for the ARM we got numbers for 10^5 signatures (cuz 10^7 takes too long)...

Thanks again for considering the improved implementation for the pqm4 library!!!

**Pull request merged**